crates/cext/src/transpiler/target.rs

```diff
@@ -0,0 +1,1296 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2025
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use crate::exit_codes::ExitCode;
+use crate::pointers::{const_ptr_as_ref, mut_ptr_as_ref};
+use indexmap::IndexMap;
+use qiskit_accelerate::{
+    nlayout::PhysicalQubit,
+    target_transpiler::{InstructionProperties, Qargs, Target},
+};
+use qiskit_circuit::operations::{Operation, Param, StandardGate};
+use std::ffi::{c_char, CStr, CString};
+use std::mem::forget;
+use std::ptr::null_mut;
+
+/// @ingroup QkTarget
+/// Construct a new Target with the given number of qubits.
+///
+/// @param description A string describing the Target, or a NULL pointer.
+/// @param num_qubits The number of qubits the Target will explicitly support
+///     or a NULL pointer to dynamically assign a number.
+/// @param dt The dt value for the system time resolution of input, or
+///     a NULL pointer.
+/// @param granularity The value for the minimum pulse gate resolution in
+///     units of ``dt``, or a NULL pointer.
+/// @param min_length minimum pulse gate length value in units of ``dt`` or a NULL
+///     pointer.
+/// @param pulse_alignment value representing a time resolution of gate
+///     instruction starting time or a NULL pointer.
+/// @param acquire_alignment value representing a time resolution of measure instruction
+///     starting time, or a NULL pointer.
+///
+/// @return A pointer to the new Target
+///
+/// # Example
+///     
+///     QkTarget *target = qk_target_new("New Target", 5, NULL, NULL, NULL, NULL, NULL);
+///
+/// # Safety
+///
+/// All arguments will be either valid pointers or NULL pointers.
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_target_new(
+    description: *mut c_char,
+    num_qubits: *const usize,
+    dt: *const f64,
+    granularity: *const u32,
+    min_length: *const usize,
+    pulse_alignment: *const u32,
+    acquire_alignment: *const u32,
```

Comment:
This is a really weird pattern for these arguments. I get that you're trying to have `Option<T>` here by using the null pointer, but treating these all as pointers is not a good way to do this. Instead I would just have methods to set these. I don't think `num_qubits` being `None` is something we need in C, it's a weird edge case that doesn't exist in real targets. So I would make this signature `qk_target_new(num_qubits: u32)`. Then add a bunch of new functions for `qk_target_set_dt(target: *mut Target, dt: f64)`, etc.